lib/web_ui/lib/src/engine/dom_canvas.dart

```diff
@@ -204,18 +204,25 @@ void _applyRRectBorderRadius(html.CssStyleDeclaration style, ui.RRect rrect) {
       rrect.trRadiusX == rrect.trRadiusY &&
       rrect.blRadiusX == rrect.blRadiusY &&
       rrect.brRadiusX == rrect.brRadiusY) {
-    style.borderRadius = '${rrect.blRadiusX.toStringAsFixed(3)}px';
+    style.borderRadius = '${_borderStrokeToCssUnit(rrect.blRadiusX)}';
     return;
   }
   // Non-uniform. Apply each corner radius.
-  style.borderTopLeftRadius = '${rrect.tlRadiusX.toStringAsFixed(3)}px '
-      '${rrect.tlRadiusY.toStringAsFixed(3)}px';
-  style.borderTopRightRadius = '${rrect.trRadiusX.toStringAsFixed(3)}px '
-      '${rrect.trRadiusY.toStringAsFixed(3)}px';
-  style.borderBottomLeftRadius = '${rrect.blRadiusX.toStringAsFixed(3)}px '
-      '${rrect.blRadiusY.toStringAsFixed(3)}px';
-  style.borderBottomRightRadius = '${rrect.brRadiusX.toStringAsFixed(3)}px '
-      '${rrect.brRadiusY.toStringAsFixed(3)}px';
+  style.borderTopLeftRadius = '${_borderStrokeToCssUnit(rrect.tlRadiusX)} '
+      '${_borderStrokeToCssUnit(rrect.tlRadiusY)}';
+  style.borderTopRightRadius = '${_borderStrokeToCssUnit(rrect.trRadiusX)} '
+      '${_borderStrokeToCssUnit(rrect.trRadiusY)}';
+  style.borderBottomLeftRadius = '${_borderStrokeToCssUnit(rrect.blRadiusX)} '
+      '${_borderStrokeToCssUnit(rrect.blRadiusY)}';
+  style.borderBottomRightRadius = '${_borderStrokeToCssUnit(rrect.brRadiusX)} '
+      '${_borderStrokeToCssUnit(rrect.brRadiusY)}';
+}
+
+String _borderStrokeToCssUnit(double value) {
+  if (value == 0) {
+    value = 1.0;
```

Comment:
Is this trying to draw a hairline? If so, does it need to be divided by DPR?

Reply:
Added // TODO: hairline nees to take into account both dpi and density.